### PR TITLE
Controller test cases Added

### PR DESCRIPTION
**Pull request by Harimaharajan**

### CODE REVIEW

File: OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs [hunk 1]
```diff
@@ -1,48 +1,55 @@
...
+            var client = server.CreateClient();
 
-            Assert.Equal((int)HttpStatusCode.OK, actualResult.StatusCode);
+            var response = await client.GetAsync(url);
+
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+        }
+
+        [Theory]
+        [InlineData("/api/SearchFlights/Search?startLocation=BUD&endDestination=LTN&departureDate=2018-12-06")]
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs]: maintain the endpoints as constants

File: OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs [hunk 2]
```diff
@@ -1,48 +1,55 @@
...
+        public async Task SearchController_InValidActionRequest_ReturnsHttpStatusNotFound404(string url)
+        {
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs:38]: InvalidActionRequest? what is the method testing?

File: OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs [hunk 3]
```diff
@@ -1,48 +1,55 @@
...
+        [InlineData("/api/SearchFlights/SearchFlightDetails?startLocation=BUD&endDestination=XYZ&departureDate=2018-12-07")]
+        public async Task SearchController_InValidSearchParameterRequest_ThrowsValidationException(string url)
         {
-            var result = searchFlightsController.SearchFlight(startLocation, null, DateTime.Now.AddDays(1)) as OkObjectResult;
+            var client = server.CreateClient();
+
+            var response = await Assert.ThrowsAsync<ValidationException>(() => client.GetAsync(url));
 
-            Assert.Equal((int)HttpStatusCode.BadRequest, result.StatusCode);
+            Assert.IsType<ValidationException>(response);
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs]: type check already done in the above statement. Check the data?

File: OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs [hunk 4]
```diff
@@ -1,48 +1,55 @@
...
         [Theory]
-        [InlineData("BUD", "LTN")]
-        public void SearchFlight_IfDestinationNullOrEmpty_ThrowsValidationException(string startLocation, string endLocation)
+        [InlineData("/api/SearchFlights/SearchFlightDetails?startLocation=BUD&endDestination=XYZ&departureDate=2018-12-07")]
+        public async Task SearchController_InValidSearchParameterRequest_ThrowsValidationException(string url)
         {
-            var result = searchFlightsController.SearchFlight(startLocation, null, DateTime.Now.AddDays(1)) as OkObjectResult;
+            var client = server.CreateClient();
+
+            var response = await Assert.ThrowsAsync<ValidationException>(() => client.GetAsync(url));
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs]: How is GetAsync throwing ValidationException!?

File: OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs (hunk 1 above)
Comment (ghost) [reply OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs]: easier way would be jus taking the data and forming the url inside

File: OnlineFlightSearchAPI/Startup.cs [hunk 5]
```diff
@@ -24,6 +25,7 @@ public void ConfigureServices(IServiceCollection services)
             services.AddScoped<ISearchFlightService, FlightService>();
             services.AddScoped<IAirportServices, AirportServices>();
             services.AddScoped<IFlightRepository, FlightRepository>();
+            services.AddScoped<IAirportRepository, AirportRepository>();
```
Comment (ghost) [inline OnlineFlightSearchAPI/Startup.cs]: need scoped?
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-2.2#service-lifetimes

**Review by ghost — CHANGES_REQUESTED** [review]
Few things missed out in the last PR since you merged it before i approved it:
Remove hardcoded data from repository classes!
I'd prefer dependency resolution inside the test methods, not at constructor level
Properly test the services, with mocked dependencies!

Do not merge with out approval!

I'd assume this PR is only for controller tests not the implementation?

File: OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs (hunk 1 above)
Comment (Harimaharajan) [reply OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs]: Made the Endpoints as constants as specified

File: OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs [hunk 6]
```diff
@@ -1,48 +1,58 @@
...
-            searchFlightsController = serviceProvider.GetService<SearchFlightsController>();
+            server = new TestServer(new WebHostBuilder()
+                .UseStartup<Startup>());
         }
 
         [Theory]
-        [InlineData("BUD", "LTN")]
-        public void SearchFlight_IfAllSearchParametersAreValid_ReturnsStatusCodeOK200(string startLocation, string endLocation)
+        [InlineData("BUD", "LTN", "2018-12-12")]
+        public async Task SearchController_ValidRequest_ReturnsHttpStatusOK200(string startLocation, string destination, string departureDate)
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs]: Unit under test is SearchController or SearchFlightDetails method?

File: OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs [hunk 7]
```diff
@@ -1,48 +1,58 @@
...
         {
-            var actualResult = searchFlightsController.SearchFlight(startLocation, endLocation, DateTime.Now.AddDays(1)) as OkObjectResult;
+            var client = server.CreateClient();
+            var url = $"/api/SearchFlights/SearchFlightDetails?startLocation={startLocation}&endDestination={destination}&departureDate={departureDate}";
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs]: could the endpoint be better?
From consumer perspective it makes no sense...

File: OnlineFlightSearchAPITestCases/Mocks/AirportRepositoryMock.cs [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+            return this;
+        }
+
+        private List<AirportDetail> GetAirportDetails()
+        {
+            List<AirportDetail> airportDetails = new List<AirportDetail>();
+            AirportDetail airportBUD = new AirportDetail("BUD", "Budapest", "Hungary", DateTime.UtcNow.AddHours(1).Kind);
+            AirportDetail airportLTN = new AirportDetail("LTN", "London Luton", "UK", DateTime.UtcNow.Kind);
+            AirportDetail airportIAD = new AirportDetail("IAD", "Washington", "USA", DateTime.UtcNow.AddHours(-5).Kind);
+            airportDetails.Add(airportBUD);
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/Mocks/AirportRepositoryMock.cs]: avoid single use variables, make them inline.

File: OnlineFlightSearchAPITestCases/Mocks/AirportServiceMock.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+﻿using Moq;
+using OnlineFlightSearchAPI.FlightServices;
+using OnlineFlightSearchAPI.Repositories;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace OnlineFlightSearchAPI.UnitTests.Mocks
+{
+    class AirportServiceMock : Mock<IAirportServices>
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/Mocks/AirportServiceMock.cs]: access modifier?

File: OnlineFlightSearchAPITestCases/Mocks/AirportServiceMock.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        public AirportServiceMock SetUp(IAirportRepository airportRerpository)
+        {
+            var mockAirportRepository = new AirportRepositoryMock().SetUp();
+
+            var mockAirportServices = new AirportServices(airportRerpository);
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/Mocks/AirportServiceMock.cs]: where is it used?

File: OnlineFlightSearchAPITestCases/Mocks/FlightRepositoryMock.cs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            FlightDetail flight3 = new FlightDetail("CM 2003", "BUD", "LTN", "2018-12-13 18:00", "3:35", 80, 10);
+            FlightDetail flight4 = new FlightDetail("CM 2004", "LTN", "IAD", "2018-12-12 06:00", "2:05", 200, 2);
+            FlightDetail flight5 = new FlightDetail("CM 2005", "LTN", "BUD", "2018-12-12 05:00", "2:00", 75, 60);
+            FlightDetail flight6 = new FlightDetail("CM 2006", "LTN", "IAD", "2018-12-12 08:00", "12:35", 45, 20);
+            FlightDetail flight7 = new FlightDetail("CM 2007", "BUD", "LTN", "2018-12-12 13:00", "12:30", 40, 33);
+            FlightDetail flight8 = new FlightDetail("CM 2008", "LTN", "BUD", "2018-12-11 09:00", "2:05", 78, 0);
+            FlightDetail flight9 = new FlightDetail("CM 2009", "BUD", "LTN", "2018-12-11 10:00", "2:00", 90, 0);
+            FlightDetail flight10 = new FlightDetail("CM 2010", "IAD", "LTN", "2018-12-11 2:00", "14:35", 190, 1);
+
+            flightDetails.Add(flight1);
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/Mocks/FlightRepositoryMock.cs]: single use variables

File: OnlineFlightSearchAPITestCases/ServiceTests/FlightServicesTests/FlightServiceTests.cs [hunk 12]
```diff
@@ -1,36 +1,48 @@
...
+using System.Linq;
 using Xunit;
 
 namespace OnlineFlightSearchAPITestCases
 {
-    public class SearchFlightTests
+    public class FlightServiceTests
     {
-        private readonly ISearchFlightService searchFlightService;
+        private ISearchFlightService searchFlightService;
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/ServiceTests/FlightServicesTests/FlightServiceTests.cs]: remove globals

File: OnlineFlightSearchAPITestCases/ServiceTests/FlightServicesTests/FlightServiceTests.cs [hunk 13]
```diff
@@ -1,36 +1,48 @@
...
 
-        public SearchFlightTests()
+        private readonly IAirportServices airportService;
+
+        private readonly IAirportRepository airportRepository;
+
+        private FlightRepositoryMock flightRepositoryMocking()
```
Comment (ghost) [inline OnlineFlightSearchAPITestCases/ServiceTests/FlightServicesTests/FlightServiceTests.cs]: method name could be better?

**Review by ghost — CHANGES_REQUESTED** [review]

File: OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs (hunk 6 above)
Comment (Harimaharajan) [reply OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs]: Replaced the Method under Test

File: OnlineFlightSearchAPITestCases/Mocks/AirportRepositoryMock.cs (hunk 8 above)
Comment (Harimaharajan) [reply OnlineFlightSearchAPITestCases/Mocks/AirportRepositoryMock.cs]: Made the objects inline

File: OnlineFlightSearchAPITestCases/Mocks/FlightRepositoryMock.cs (hunk 11 above)
Comment (Harimaharajan) [reply OnlineFlightSearchAPITestCases/Mocks/FlightRepositoryMock.cs]: made the objects inline

File: OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs (hunk 2 above)
Comment (Harimaharajan) [reply OnlineFlightSearchAPITestCases/ControllerTests/SearchFlightsControllerTests.cs:38]: Renamed the method

File: OnlineFlightSearchAPI/Startup.cs (hunk 5 above)
Comment (Harimaharajan) [reply OnlineFlightSearchAPI/Startup.cs]: Removed Dependency Injection and used Mocking

File: OnlineFlightSearchAPITestCases/Mocks/AirportServiceMock.cs (hunk 9 above)
Comment (Harimaharajan) [reply OnlineFlightSearchAPITestCases/Mocks/AirportServiceMock.cs]: Removed the Separate Mocking Class as specified

File: OnlineFlightSearchAPITestCases/Mocks/AirportServiceMock.cs (hunk 10 above)
Comment (Harimaharajan) [reply OnlineFlightSearchAPITestCases/Mocks/AirportServiceMock.cs]: Removed the Separate Mocking Class

File: OnlineFlightSearchAPITestCases/ServiceTests/FlightServicesTests/FlightServiceTests.cs (hunk 12 above)
Comment (Harimaharajan) [reply OnlineFlightSearchAPITestCases/ServiceTests/FlightServicesTests/FlightServiceTests.cs]: Removed Globals

File: OnlineFlightSearchAPITestCases/ServiceTests/FlightServicesTests/FlightServiceTests.cs (hunk 13 above)
Comment (Harimaharajan) [reply OnlineFlightSearchAPITestCases/ServiceTests/FlightServicesTests/FlightServiceTests.cs]: Removed Separate Mocking in Methods. Moved the mocking to method under test

**Comment by Harimaharajan** [comment]
Will create a new PR

**Pull request closed without merging**